sis_provisioner/dao/student.py

```diff
@@ -0,0 +1,147 @@
+
+    def _get_from_dict(self, current_dict, path: list):
+        for count, key in enumerate(path[:-1]):
+            if '[]' in key:
+                key = key.replace('[]', '')
+                return [self._get_from_dict(item, path[count+1:]) \
+                        for item in current_dict[key]]
+            elif re.search('\[\d+\]', key):
+                index = re.findall('\[\d+\]', key)[0]
+                current_dict = current_dict[key.replace(index, '')]
+                if current_dict:
+                    current_dict = current_dict[int(re.findall('\d+', index)[0])]
+                else:
+                    return None
+            else:
+                current_dict = current_dict[key]
+        return current_dict[path[-1]]
+
+    def create_handshake_dicts(self, students: list, fieldnames) -> list:
+        handshake_dicts = []
+        for student in students:
+            handshake_dict = self._init_dict(fieldnames)
+            for handshake_key, db_key in self.get_handshake_fieldnames_map().items():
+                val = self._get_from_dict(student, db_key.split(':'))
+                self._fill_dict(handshake_dict, val, [handshake_key])
+            handshake_dicts.append(handshake_dict)
+        # filter out campus 1 and college E majors
+        self._filter_students(handshake_dicts)
+        return handshake_dicts
+
+    def get_students_for_handshake(self) -> list:
+        students = self.client.get_registered_students(
+            include_employee=False,
+            include_student_transcripts=False,
+            include_student_transfers=False,
+            include_student_sports=False,
+            include_student_advisers=False,
+            include_student_intended_majors=False,
+            include_student_pending_majors=False,
+            include_student_requested_majors=False)
+
+        handshake_students = []
+
+        for person in students:
+            if not self.valid_class_code(person.student.class_code):
+                continue
+
+            if not self.valid_campus_code(person.student.campus_code):
+                continue
+
+            if not self.valid_major_code(person.student.majors):
+                continue
+
+            self._update_college_for_majors(person.student.majors)
+
+            handshake_students.append(person)
+
+        return handshake_students
+
+    def get_student_csv(self, filename) -> str:
+        students = self.get_students_for_handshake()
+        students_list = [student.to_dict() for student in students]
+        fieldnames = [key.replace('[]', '') for key in self.get_handshake_fieldnames_map().keys()]
+        students_dicts = self.create_handshake_dicts(students_list, fieldnames)
+
+        if students_dicts:
+            with open(filename, 'w', newline='') as csvfile:
+                csv.register_dialect('unix_newline', lineterminator='\n')
+                writer = csv.DictWriter(csvfile, fieldnames, dialect='unix_newline')
```

Comment:
It was easier to use DictWriter now since the processing is mainly done in create_handshake_dicts